aries_vcx/src/common/credentials/encoding.rs

```diff
@@ -1,99 +1,140 @@
 use crate::utils::openssl::encode;
-use std::collections::HashMap;
 
-pub fn encode_attributes(attributes: &str) -> VcxResult<String> {
-    let mut dictionary = HashMap::new();
-    match serde_json::from_str::<HashMap<String, serde_json::Value>>(attributes) {
-        Ok(mut attributes) => {
-            for (attr, attr_data) in attributes.iter_mut() {
-                let first_attr = match attr_data {
-                    // new style input such as {"address2":"101 Wilson Lane"}
-                    serde_json::Value::String(str_type) => str_type,
-
-                    // old style input such as {"address2":["101 Wilson Lane"]}
-                    serde_json::Value::Array(array_type) => {
-                        let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
-                            Some(x) => x,
-                            None => {
-                                return Err(AriesVcxError::from_msg(
-                                    AriesVcxErrorKind::InvalidAttributesStructure,
-                                    "Attribute value not found",
-                                ));
-                            }
-                        };
-
-                        warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
-                        attrib_value
-                    }
-                    _ => {
-                        return Err(AriesVcxError::from_msg(
-                            AriesVcxErrorKind::InvalidJson,
-                            "Invalid Json for Attribute data",
-                        ));
-                    }
-                };
-
-                let encoded = encode(first_attr)?;
-                let attrib_values = json!({
-                    "raw": first_attr,
-                    "encoded": encoded
-                });
-
-                dictionary.insert(attr.to_string(), attrib_values);
+/// `CredentialAttributeData` represents contains all the attributes data
+/// in a credential.
+#[derive(Debug, Clone)]
+pub struct CredentialAttributeData<D> {
+    /// A reference to the attributes data in the credentials.
+    data: D,
+}
+
+/// `EncodedCredentialAttributes` contains the encoded credential attributes data.
+#[derive(Debug, Clone)]
+pub struct EncodedCredentialAttributes {
+    /// Encoded data text.
+    pub data: String,
+}
+
+impl<D> CredentialAttributeData<D>
+where
+    D: AsRef<str>,
+{
+    /// Create new attributes data.
+    pub fn new(data: D) -> Self {
+        Self { data }
+    }
+
+    /// Encodes the attributes in a credential.
+    pub fn encode(&self) -> VcxResult<EncodedCredentialAttributes> {
+        let mut dictionary = HashMap::new();
+        match serde_json::from_str::<HashMap<String, serde_json::Value>>(self.data.as_ref()) {
+            Ok(attributes) => Self::encode_table_attribute(attributes, &mut dictionary),
```

Comment:
Rather than passing mutable dictionary in, you can allocate it in the encoding functions and return it, it's cleaner